### PR TITLE
fix(kubernetes): No dups when extracting images

### PR DESCRIPTION
**Pull request by bo156**
**By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

[//]: # "
    # PR Title
    Be aware that we use the title to create changelog automatically and therefore only allow specific prefixes
    - break:    to indicate a breaking change, this supersedes any of the types
    - feat:     to indicate new features or checks
    - fix:      to indicate a bugfix or handling of edge cases of existing checks
    - docs:     to indicate an update to our documentation
    - chore:    to indicate adjustments to workflow files or dependency updates
    - platform: to indicate a change needed for the platform
    Additionally a scope is needs to be added to the prefix, which indicates the targeted framework, in doubt choose 'general'.
    #    
    Allowed prefixs:
    ansible|argo|arm|azure|bicep|bitbucket|circleci|cloudformation|dockerfile|github|gha|gitlab|helm|kubernetes|kustomize|openapi|sca|secrets|serverless|terraform|general|graph|terraform_plan|terraform_json
    #
    ex.
    feat(terraform): add CKV_AWS_123 to ensure that VPC Endpoint Service is configured for Manual Acceptance
"

## Description

In some cases a k8s configuration might contain images names more than once (for example - using the same image as `initContainer` and `container`). 
This PR makes sure we don't return any duplications while trying to extract resources from the template

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my feature, policy, or fix is effective and works
- [ ] New and existing tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: checkov/kubernetes/image_referencer/provider/k8s.py [hunk 1]
```diff
@@ -90,6 +54,20 @@ def extract_images_from_containers(containers: Any) -> list[str]:
     return image_names
 
 
+def _extract_images_from_spec(spec: dict[str, Any]) -> list[str]:
+    image_names: list[str] = []
```
Comment (gruebel) [inline checkov/kubernetes/image_referencer/provider/k8s.py:58]: how about making this a `set()` then you don't need to transform it to a set and back to list 🙂

Comment (bo156) [reply checkov/kubernetes/image_referencer/provider/k8s.py]: good idea - fixed

File: checkov/kubernetes/image_referencer/provider/k8s.py [hunk 2]
```diff
@@ -19,77 +19,55 @@ def __init__(self, graph_connector: DiGraph) -> None:
...
             if isinstance(container, dict):
                 image = container.get("image")
                 if image and isinstance(image, str):
-                    image_names.append(image)
+                    image_names.add(image)
 
     return image_names
 
 
+def _extract_images_from_spec(spec: dict[str, Any] | None) -> list[str]:
```
Comment (omryMen) [inline checkov/kubernetes/image_referencer/provider/k8s.py:57]: Nice !

File: checkov/kubernetes/image_referencer/provider/k8s.py [hunk 3]
```diff
@@ -19,77 +19,55 @@ def __init__(self, graph_connector: DiGraph) -> None:
...
+    image_names: set[str] = set()
+
+    if isinstance(spec, dict):
+        containers = spec.get("containers")
+        image_names.update(extract_images_from_containers(containers=containers))
+
+        containers = spec.get("initContainers")
+        image_names.update(extract_images_from_containers(containers=containers))
```
Comment (omryMen) [inline checkov/kubernetes/image_referencer/provider/k8s.py:65]: maybe update only if containers is not None ?

File: checkov/kubernetes/image_referencer/provider/k8s.py (hunk 1 above)
Comment (omryMen) [reply checkov/kubernetes/image_referencer/provider/k8s.py]: maybe all of the functions here can return sets ?

Comment (gruebel) [reply checkov/kubernetes/image_referencer/provider/k8s.py]: also thought the same, but I think it is a bigger change, because of the typing mismatch.

Comment (bo156) [reply checkov/kubernetes/image_referencer/provider/k8s.py]: Just didn't watch to change api

File: checkov/kubernetes/image_referencer/provider/k8s.py (hunk 3 above)
Comment (bo156) [reply checkov/kubernetes/image_referencer/provider/k8s.py:65]: it just adds some more code and makes it look more ugly :/

**Review by omryMen — APPROVED** [review]

**Review by gruebel — APPROVED** [review]
🏖️

**Pull request merged**